python/ray/tune/automl/search_policy.py

```diff
@@ -0,0 +1,229 @@
+
+    def next_trials(self):
+        if self._undone_count > 0:
+            # Last round not finished
+            return []
+
+        trials = []
+        raw_param_list, extra_arg_list = self._select()
+        if not extra_arg_list:
+            extra_arg_list = [None] * len(raw_param_list)
+
+        for param_config, extra_arg in zip(raw_param_list, extra_arg_list):
+            tag = ''
+            new_spec = copy.deepcopy(self.experiment.spec)
+            for path, value in param_config.items():
+                tag += '%s=%s-' % (path.split('.')[-1], value)
+                merge_config(path, value, new_spec['config'])
+
+            trial = create_trial_from_spec(
+                new_spec, self.experiment.name,
+                self._parser, experiment_tag=tag
+            )
+
+            # AutoML specific fields set in Trial
+            trial.results = []
+            trial.best_result = None
+            trial.param_config = param_config
+            trial.extra_arg = extra_arg
+
+            trials.append(trial)
+            self._running_trials[trial.trial_id] = trial
+            # self.summary.on_trial_create(trial)
+
+        ntrial = len(trials)
+        self._iteration += 1
+        self._undone_count = ntrial
+        self._total_trial_num += ntrial
+        self._start_ts = time.time()
+        logging.info(("=========== BEGIN Experiment-Round: %s "
+                      "[%s NEW | %s TOTAL] ===========")
+                     % (self._iteration, ntrial, self._total_trial_num))
+        return trials
+
+    def _trial_add_result(self, trial, result):
+        trial.results.append(result)
+        if trial.best_result is None \
+                or result[self.reward_attr] > trial.best_result[self.reward_attr]:
+            trial.best_result = result
+
+    def on_trial_result(self, trial_id, result):
+        if not result:
+            return
+
+        # Update trial's best result
+        trial = self._running_trials[trial_id]
+        self._trial_add_result(trial, result)
+        self.summary.on_trial_result(trial, result)
+
+        # Update job's best trial
+        if self.best_trial is None \
+                or (result[self.reward_attr]
+                    > self.best_trial.best_result[self.reward_attr]):
+            self.best_trial = self._running_trials[trial_id]
+
+    def on_trial_complete(self,
+                          trial_id,
+                          result=None,
+                          error=False,
+                          early_terminated=False):
+        self.on_trial_result(trial_id, result)
+        self._undone_count -= 1
+        if self._undone_count == 0:
+            total = len(self._running_trials)
+            succ = len(filter(lambda t: t.status == Trial.TERMINATED,
+                              self._running_trials.values()))
+            this_trial = self._running_trials[trial_id]
+            if this_trial.status == Trial.RUNNING and not error:
```

Comment:
this should be if not errored and if not early_terminated right?

Reply:
We regard early terminated trials as successful ones to keep consistence with `TrialRunner`, where the `Trial.status` is `TERMINATED` instead of `ERROR` if early terminated. I think the early_terminate flag should be kept in Trial, thus the upper layer could know more about the trial.